### ISSUE TITLE
limit_posts is unusable with post_url

### ISSUE THREAD

**Issue by jmmv** [jekyll/jekyll#5890]
- [X] I believe this to be a bug, not a question about using Jekyll.
- [X] I updated to the latest Jekyll (or) if on GitHub Pages to the latest `github-pages`
- [X] I read the CONTRIBUTION file at https://jekyllrb.com/docs/contributing/
- [ ] This is a feature request.

---

- [X] I am on (or have tested on) ***macOS*** 10+
- [ ] I am on (or have tested on) ***Debian/Ubuntu*** GNU/Linux
- [ ] I am on (or have tested on) ***Fedora*** GNU/Linux
- [ ] I am on (or have tested on) ***Arch*** GNU/Linux
- [ ] I am on (or have tested on) ***Other*** GNU/Linux
- [ ] I am on (or have tested on) ***Windows*** 10+

---

- [ ] I was trying to install.
- [ ] There is a broken Plugin API.
- [ ] I had an error on GitHub Pages, and I have reproduced it locally.
- [ ] I had an error on GitHub Pages, and GitHub Support said it was a Jekyll Bug.
- [ ] I had an error on GitHub Pages and I did not test it locally.
- [X] I was trying to build.
- [ ] It was another bug.

## My Reproduction Steps

I have a jekyll-based blog in which I imported years of historical posts from a different blog. I currently have 600+ posts and building the whole site to preview changes to a single post is painful: the whole build takes over a minute, and I don't care about building old posts for drafting purposes.

For this reason, I am using the `--limit_posts` flag to build a "dev version" of my site that only includes the latest posts. This lets me quickly iterate on changes/drafts.

The problem is that my posts are heavily cross-referenced with `{%post_url%}`. If a recent post references an older post in this manner, and the old post is skipped because `--limit_posts` does not include it in the build, the build breaks saying that the `post_url` reference is broken.

An easy reproduction process is:

1. Create a blog with two posts.
2. Make the newest blog post reference the oldest one using `post_url`.
3. Build the blog with `--limit_posts=1`, and see it fail.

This effectively makes `limit_posts` useless.

## The Output I Wanted

I would like invalid `post_url` links to be ignored when `--limit_posts` is used, possibly logging to the console.

Alternatively, I'd like a flag to suppress `post_url` errors. This makes sense if `--limit_posts` is intended to be usable for "final production sites" instead of just development reasons.

**Comment by parkr** [comment]
> I would like invalid post_url links to be ignored when --limit_posts is used, possibly logging to the console.
> 
> Alternatively, I'd like a flag to suppress post_url errors. This makes sense if --limit_posts is intended to be usable for "final production sites" instead of just development reasons.

Neither of these solutions will work, as both provide less-than-optimal behaviour. If you're trying to link to something that doesn't exist, Jekyll will be unable to build your site.

A solution I would accept is to have `limit_posts` simply limit how many items from `site.posts` are *written*. That would mean everything is read and rendered, but only some are written. Would you please submit a PR for this? We can continue chatting about it there.

**Comment by jmmv** [comment]
You could make `post_url` emit a link to the current post to keep Jekyll happy. Or I'm sure there are other alternatives that could be pursued to make something work.

Note that rendering the whole site and then only writing some entries out does not help at all for my use case: what I'm trying to do is *cut down the build time* of the site for faster iteration on new drafts / site redesigns, not to generate a small set of posts. Thus I do really want to avoid processing the vast majority of the posts. As I said, it takes 1 minute to build my site at the moment, and limiting the posts brings the build time down to less than 10 seconds.

What do you want me to submit a PR for? I don't have any code changes for this, and I don't know Ruby... so asking for a PR and closing this bug basically means that this request will be forgotten.

**Comment by parkr** [comment]
Hey @jmmv! Closing an issue does not mean that it won't get a response. This is a community-run project and one-off pull requests from users are the way we get most of our features built (especially from folks who have never used Ruby before, believe it or not!), so that's why I asked for a PR. I closed the issue because your proposal is to essentially fail quietly which is strictly against our philosophy of failing loudly. No offense was meant!

If you don't need your entire site to be built, then I'd recommend moving all but 1 of your posts from `_posts` to `_not_posts` or some other subdirectory. If you can select posts based on the absence of the `post_url` string (e.g. `grep -v post_url`) then keep the first 10, you'd be in business. What I usually do when I am redesigning a site is pare down the number of items that have to be rendered, but keep a subset that are representative of the content I have. I move all the files I don't need to a subdirectory that starts with an underscore (e.g. `_junk`) and put it back once I'm happy with the design. This is a a fairly common way of doing things. In fact, it's built in to [the Octopress 2 workflow](https://github.com/imathis/octopress/blob/5717a50f4e4964ea8423997d6456b04c653f0348/Rakefile#L161-L169). Would that work for you here?

> it takes 1 minute to build my site at the moment, and limiting the posts brings the build time down to less than 10 seconds.

Hm, I wonder what's going on there. Running `jekyll build --profile` might help get that down even further (liquid profiling). A site with a handful of posts taking more than a second or two might have some other places for optimization.

**Comment by jmmv** [comment]
Finally got a chance to run my site with `--profile` and I found some good things for improvement. Could cut down full site builds from 60+ seconds to 8 seconds so I really don't need this feature any longer.

But the site is still taking 8 seconds because there are ~700 posts, and now most of the time is spent processing common files over and over again without obvious things I could remove.

I noticed that Jekyll barely consumes one CPU, so I'm wondering if it's threaded at all. At this point, parallel generation of posts is the only thing that'd help me I think, but I have no idea how hard that'd be to implement.
